src/lib.rs

```diff
@@ -127,181 +44,191 @@ macro_rules! symbol {
     let mut new_subs = vec![];
+
     for substitution in all_subs {
-        let down_to_earth_atom = substitute(atom, substitution);
-        for entry in kb.iter() {
+        let down_to_earth_atom = substitution.apply_to_atom(atom);
+
+        for entry in kb.atoms.iter() {
             if let Some(extension) = unify(&down_to_earth_atom, entry) {
-                new_subs.push(Substitution(
-                    substitution
-                        .iter()
-                        .cloned()
-                        .chain(extension.0.into_iter())
-                        .collect(),
-                ));
+                let mut new_map = substitution.mapping.clone();
+                new_map.extend(extension.mapping.into_iter());
+                new_subs.push(Substitution { mapping: new_map });
             }
         }
     }
+
     new_subs
 }
 
-pub fn unify(a: &Atom, b: &Atom) -> Option<Substitution> {
+fn unify(a: &Atom, b: &Atom) -> Option<Substitution> {
     if a.pred_sym != b.pred_sym || a.terms.len() != b.terms.len() {
         return None;
     }
+
     let mut subs = Substitution::default();
-    for pair in a.terms.iter().zip(b.terms.iter()) {
-        match pair {
-            (_, Term::Var(_)) => panic!("The second atom is assumed to be ground"),
-            (v @ Term::Var(_), s @ Term::Sym(_)) => match subs.lookup(v) {
-                Some(s2 @ Term::Sym(_)) if s2 != s => {
-                    return None;
+    for (a_term, b_term) in a.terms.iter().zip(b.terms.iter()) {
+        if matches!(b_term, Term::Var(_)) {
+            panic!("The second atom is assumed to be ground")
+        }
+
+        match *a_term {
+            Term::Var(_) => {
+                let substituted = subs.lookup(a_term);
+                match substituted {
+                    Some(a_term @ Term::Sym(_)) if a_term != b_term => {
+                        return None;
+                    }
+                    _ => {
+                        subs.mapping.insert(a_term.clone(), b_term.clone());
+                    }
                 }
-                _ => subs.push((v.clone(), s.clone())),
-            },
-            (s1 @ Term::Sym(_), s2 @ Term::Sym(_)) if s1 != s2 => {
+            }
+            Term::Sym(_) if a_term != b_term => {
                 return None;
             }
             _ => {}
         }
     }
-    Some(subs)
-}
 
-pub fn substitute(atom: &Atom, substitution: &Substitution) -> Atom {
-    let mut atom = atom.clone();
-    let terms = atom
-        .terms
-        .iter()
-        .map(|term| match term {
-            v @ Term::Var(_) => substitution.lookup(v).cloned().unwrap_or_else(|| v.clone()),
-            Term::Sym(_) => term.clone(),
-        })
-        .collect();
-    atom.terms = terms;
-    atom
+    Some(subs)
 }
 
 #[cfg(test)]
 mod tests {
-    use alloc::{format, string::ToString as _};
+    use alloc::string::ToString as _;
 
-    use super::*;
+    use super::{parsing::parse, *};
 
     #[test]
```

Comment:
Do these tests need to feature gate on parsing now though?

```
#[cfg(feature = "parsing")]
```

Reply:
nope, `test` is like a feature itself that's enabled only when `cargo test` runs